perception/front_vehicle_velocity_estimator/src/front_vehicle_velocity_estimator_node/front_vehicle_velocity_estimator_node.cpp

```diff
@@ -0,0 +1,200 @@
+#include <memory>
+#include <string>
+#include <vector>
+
+using namespace std::literals;
+using std::chrono::duration;
+using std::chrono::duration_cast;
+using std::chrono::nanoseconds;
+
+namespace
+{
+template <class T>
+bool update_param(
+  const std::vector<rclcpp::Parameter> & params, const std::string & name, T & value)
+{
+  const auto itr = std::find_if(
+    params.cbegin(), params.cend(),
+    [&name](const rclcpp::Parameter & p) { return p.get_name() == name; });
+
+  // Not found
+  if (itr == params.cend()) {
+    return false;
+  }
+
+  value = itr->template get_value<T>();
+  return true;
+}
+}  // namespace
+
+namespace front_vehicle_velocity_estimator
+{
+
+FrontVehicleVelocityEstimatorNode::FrontVehicleVelocityEstimatorNode(
+  const rclcpp::NodeOptions & node_options)
+: Node("front_vehicle_velocity_estimator", node_options)
+{
+  // Parameter Server
+  set_param_res_ = this->add_on_set_parameters_callback(
+    std::bind(&FrontVehicleVelocityEstimatorNode::onSetParam, this, std::placeholders::_1));
+
+  // Node Parameter
+  node_param_.update_rate_hz = declare_parameter<double>("node_params.update_rate_hz", 10.0);
+
+  // Core Parameter
+  core_param_.moving_average_num = declare_parameter<int>("core_params.moving_average_num", 1);
+  core_param_.threshold_pointcloud_z_high =
+    declare_parameter<float>("core_params.threshold_pointcloud_z_high", 1.0f);
+  core_param_.threshold_pointcloud_z_low =
+    declare_parameter<float>("core_params.threshold_pointcloud_z_low", 0.6f);
+  core_param_.threshold_relative_velocity =
+    declare_parameter<double>("core_params.threshold_relative_velocity", 10.0);
+  core_param_.threshold_absolute_velocity =
+    declare_parameter<double>("core_params.threshold_absolute_velocity", 20.0);
+
+  // Core
+  front_vehicle_velocity_estimator_ = std::make_unique<FrontVehicleVelocityEstimator>(get_logger());
+  front_vehicle_velocity_estimator_->setParam(core_param_);
+
+  // Subscriber
+  sub_pointcloud_ = create_subscription<PointCloud2>(
+    "~/input/pointcloud", rclcpp::SensorDataQoS(),
+    std::bind(&FrontVehicleVelocityEstimatorNode::onPointcloud, this, std::placeholders::_1));
+  sub_objects_ = create_subscription<DetectedObjects>(
+    "~/input/objects", rclcpp::QoS{1},
+    std::bind(&FrontVehicleVelocityEstimatorNode::onObjects, this, std::placeholders::_1));
+  sub_odometry_ = create_subscription<Odometry>(
+    "~/input/odometry", rclcpp::QoS{1},
+    std::bind(&FrontVehicleVelocityEstimatorNode::onOdometry, this, std::placeholders::_1));
+
+  // Publisher
+  pub_objects_ = create_publisher<DetectedObjects>("~/output/objects", 1);
+  pub_nearest_neighbor_pointcloud_ =
+    create_publisher<PointCloud2>("~/debug/nearest_neighbor_pointcloud", 1);
+
+  // Timer
+  const auto update_period_ns = rclcpp::Rate(node_param_.update_rate_hz).period();
+  timer_ = rclcpp::create_timer(
+    this, get_clock(), update_period_ns,
+    std::bind(&FrontVehicleVelocityEstimatorNode::onTimer, this));
```

Comment:
I think it is better to use message_filter callback because onTimer makes a large delay.

Reply:
Add message filter

Reply:
Thank you :+1:

Reply:
If it is possible, pls write the commit id :pray:

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/1536/commits/2b7351c066725f7346f305489dd050782db79974

Reply:
Why onTimer function is needed?
I think it can be replaced onData function.

Reply:
Sorry, I forgot it.
Fixed in 
https://github.com/autowarefoundation/autoware.universe/pull/1536/commits/e13af3bc35ed0754be8346dcc8f0a964bc8cc1f5

Reply:
:+1: